src/models/partial_tx.ts

```diff
@@ -0,0 +1,449 @@
+
+    const dataArr = serialized.split('|');
+    const changeOutputs = dataArr.slice(2).map(x => +x);
+
+    if (dataArr.length < 2 || dataArr[0] !== 'PartialTx' || !changeOutputs.every(x => Number.isInteger(x))) {
+      throw new SyntaxError('Invalid PartialTx');
+    }
+
+    const txHex = dataArr[1];
+    const tx = helpers.createTxFromHex(txHex, network);
+
+    const instance = new PartialTx(network);
+
+    for (const input of tx.inputs) {
+      await instance.addInput(input.hash, input.index);
+    }
+
+    for (const [index, output] of tx.outputs.entries()) {
+      // validate script
+      const script = output.parseScript(network);
+      if (!(script instanceof P2PKH || script instanceof P2SH)) {
+        throw new UnsupportedScriptError('Unsupported script type');
+      }
+
+      if (output.isAuthority()) {
+        // TODO: we dont allow passing authority on atomic swap
+        throw new Error('Authority outputs are unsupported');
+      }
+
+      const tokenIndex = output.tokenData & TOKEN_INDEX_MASK;
+      const token = tokenIndex === 0 ? HATHOR_TOKEN_CONFIG.uid : tx.tokens[tokenIndex-1];
+      instance.addOutput(
+        output.value,
+        output.script,
+        token,
+        changeOutputs.indexOf(index) > -1,
+      );
+    }
+
+    return instance;
+
+  }
+}
+
+/**
+ * This class is meant to aggregate input data for a transaction.
+ *
+ * The `hash` is an identifier of the transaction (usually the dataToSign in hex format)
+ * this way any input data added should identify that it is from the same transaction.
+ *
+ * The input data is saved instead of the signature to allow collecting from MultiSig wallets
+ * since for an input we can have multiple signatures.
+ */
+export class PartialTxInputData {
+  data: Record<number, Buffer>;
+  hash: string;
+  inputsLen: number;
+
+  constructor(hash: string, inputsLen: number) {
+    this.data = {};
+    this.hash = hash;
+    this.inputsLen = inputsLen;
+  }
+
+  /**
+   * Add an input data to the record.
+   *
+   * @param {number} index The input index this data relates to.
+   * @param {Buffer} inputData Input data bytes.
+   *
+   * @throws {IndexOOBError} index should be inside the inputs array.
+   *
+   * @memberof PartialTxInputData
+   * @inner
+   */
+  addData(index: number, inputData: Buffer) {
+    if (index >= this.inputsLen) {
+      throw new IndexOOBError(`Index ${index} is out of bounds for the ${this.inputsLen} inputs`);
+    }
+    this.data[index] = inputData;
+  }
+
+  /**
+   * Return true if we have an input data for each input.
+   *
+   * @return {boolean}
+   * @memberof PartialTxInputData
+   * @inner
+   */
+  isComplete(): boolean {
+    return Object.values(this.data).length === this.inputsLen;
+  }
+
+  /**
+   * Serialize the current PartialTxInputData into an UTF8 string.
+   *
+   * @returns {string}
+   * @memberof PartialTxInputData
+   * @inner
+   */
+  serialize(): string {
+    const arr = ['PartialTxInputData', this.hash];
```

Comment:
I think it would be a good code pattern if the `PartialTxInputData` became a exported constant instead of a hardcoded string.

If this ever needs to be used somewhere else than this file, it will be easier to maintain, just by exporting it and keeping the value centralized.